File: src/rpc_types/blocks.py
from typing import List, Optional
from datetime import datetime, date
from pydantic import BaseModel

class Withdrawal(BaseModel):
    address: str
    amount: int
    index: int
    validator_index: int

class BaseBlock(BaseModel):
    model_config = {
        "arbitrary_types_allowed": False,
    }
    
    base_fee_per_gas: Optional[int] = None
    block_time: datetime
    block_date: date
    difficulty: int
    extra_data: Optional[str] = None
    gas_limit: int
    gas_used: int
    hash: str
    logs_bloom: str
    miner: str
    mix_hash: str
    nonce: str
    number: int
    parent_hash: str
    receipts_root: str
    sha3_uncles: str
    size: int
    state_root: str
    total_difficulty: int
    transactions: List[str] = []
    transactions_root: str
    uncles: List[str] = []

class ArbitrumBlock(BaseBlock):
    l1_block_number: int
    send_count: Optional[int] = None
    send_root: Optional[str] = None

class EthereumBlock(BaseBlock):
    blob_gas_used: Optional[int] = None
    excess_blob_gas: Optional[int] = None
    parent_beacon_block_root: Optional[str] = None
    withdrawals: Optional[List[Withdrawal]] = []
    withdrawals_root: Optional[str] = None

class ZKsyncBlock(BaseBlock):
    l1_batch_number: Optional[int] = None
    l1_batch_time: Optional[datetime] = None
    seal_fields: List[str] = []

File: rust/src/models/indexed/blocks.rs
use alloy_primitives::{Address, Bloom, Bytes, FixedBytes, TxKind, Uint};

use chrono::{DateTime, NaiveDate, Utc};

use crate::models::common::{ChainId, TransactionTo};

#[derive(Debug)]
pub struct TransformedBlockData {
    pub chain_id: ChainId,
    pub hash: FixedBytes<32>,
    pub parent_hash: FixedBytes<32>,
    pub ommers_hash: FixedBytes<32>,
    pub beneficiary: Address,
    pub state_root: FixedBytes<32>,
    pub transactions_root: FixedBytes<32>,
    pub receipts_root: FixedBytes<32>,
    pub logs_bloom: Bloom,
    pub difficulty: Uint<256, 4>,
    pub number: u64,
    pub gas_limit: u64,
    pub gas_used: u64,
    pub block_time: DateTime<Utc>,
    pub block_date: NaiveDate,
    pub extra_data: Bytes,
    pub mix_hash: FixedBytes<32>,
    pub nonce: FixedBytes<8>,
    pub base_fee_per_gas: Option<u64>,
    pub withdrawals_root: Option<FixedBytes<32>>,
    pub blob_gas_used: Option<u64>,
    pub excess_blob_gas: Option<u64>,
    pub parent_beacon_block_root: Option<FixedBytes<32>>,
    pub requests_hash: Option<FixedBytes<32>>,
    pub target_blobs_per_block: Option<u64>,
    pub total_difficulty: Option<Uint<256, 4>>,
    pub size: Option<Uint<256, 4>>,
}
